src/librbd/operation/SnapshotRollbackRequest.cc

```diff
@@ -43,6 +44,15 @@ class C_RollbackObject : public C_AsyncObjectThrottle<I> {
     ldout(cct, 20) << "C_RollbackObject: " << __func__ << ": object_num="
                    << m_object_num << dendl;
 
+    {
+      RWLock::RLocker snap_locker(image_ctx.snap_lock);
+      if (m_snap_object_map != nullptr &&
```

Comment:
Couple things: (1) you need to bounds-check the object num since you might have had an image resize, and (2) since the object map will be copied from snap to HEAD revision, you would also need to rollback (or remove) if the snapshot's object map is not set but the HEAD revision object map is set.

Reply:
@dillaman  Updated for (1), but unit tests still need to be fixed. Thank you.
I didn't get your points about (2), did you mean the image’s object map is not set when creating snapshot and set image's object map after created? like this steps:
```
[root@ ~]# rbd create img1 -s 1G --image-feature exclusive-lock
[root@ ~]# rbd snap create img1@snap1
[root@ ~]# rbd feature enable img1 object-map
```
If so, I think that after the image's object map is set, the snapshot will be set auto.
```
[root@ ~]# rbd info img1 | grep flags
        flags: object map invalid
[root@ ~]# rbd info img1@snap1 | grep flags
        flags: object map invalid
```

Reply:
@Songweibin No, I mean the following (simplified case):

image@snap: object ABC state is ```OBJECT_NONEXISTENT```
image: object ABC state is ```OBJECT_EXISTS```

In that case, you need to either delete object ABC at HEAD or roll it back to ```snap``` so that the OSD deletes it.

Reply:
@dillaman  In this case, it does not go into `if {... return 1;}`, but continues to execute the original process. Actually, my changes only work for both image@snap and image's object ABC are OBJECT_NONEXISTENT.

Reply:
OK, guess I missed the change against the HEAD object map.